<a href="https://colab.research.google.com/github/juanpalominoh/Fundamentos-de-Econometria/blob/main/Tema_5_Errores_de_Especificaci%C3%B3n_y_Teorema_Frisch_Waugh.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tema 5: Errores de especificación y Teorema Frisch-Waugh

In [ ]:
# Configurar el tiempo limite de carga 
options(timeout=3000)
getOption('timeout')

[1] 3000

El link del drive donde se encuentra el archivo es: https://drive.google.com/file/d/1WFCu_CwsWhqKjO7b1R1rdVPgGmfTwFBp/view?usp=sharing

In [ ]:
system("gdown --id 1WFCu_CwsWhqKjO7b1R1rdVPgGmfTwFBp")

In [ ]:
unzip("/content/enaho_2021.zip")

Cargamos una base procesada de la ENAHO 2021. 

In [ ]:
# Cargar el modulo 5 de la ENAHO 
library(haven)
enaho <- read_dta("/content/enaho_2021.dta")
head(enaho)

pais aÑo  mes dpto zona urban conglome vivienda hogar codperso ⋯ gruocu
1 Perú 2021 2   1    2    1     5007     003      11    01       ⋯ 4     
2 Perú 2021 2   1    2    1     5007     012      11    01       ⋯ 5     
3 Perú 2021 2   1    2    1     5007     022      11    01       ⋯ 4     
4 Perú 2021 2   1    2    1     5007     022      11    02       ⋯ 4     
5 Perú 2021 1   1    2    1     5009     041      11    01       ⋯ 5     
6 Perú 2021 1   1    2    1     5009     098      11    02       ⋯ 5     
  skill_ocu ingreso   lnwage   sis_pension informal lntamaño  fac500a   sch
1 2         6433.5938 8.769289 1           0        9.2101402  68.58327 17 
2 2         1062.1659 6.968065 0           1        1.3862944  86.84176 11 
3 2         2224.6653 7.707362 1           0        9.2101402  42.10129 16 
4 2         1619.2104 7.389694 1           0        9.2101402  33.74992 14 
5 2          397.5833 5.985404 0           1        0.6931472 222.03047  0 
6 2         1223.1825 7.109211 0           1        1.0986123 487.35910 11 
  exper
1 33   
2  4   
3 40   
4 41   
5 56   
6 26

## Estadística Descriptiva

In [ ]:
# Ver los nombres de las variables. Revisar diccionario para saber que significa cada una de ellas.
names(enaho)

[1] "pais"        "aÑo"         "mes"         "dpto"        "zona"       
 [6] "urban"       "conglome"    "vivienda"    "hogar"       "codperso"   
[11] "mujer"       "edad"        "civil"       "educ"        "jefe"       
[16] "residente"   "indigena"    "tenure"      "pea"         "peao"       
[21] "act_sec"     "sector"      "gruocu"      "skill_ocu"   "ingreso"    
[26] "lnwage"      "sis_pension" "informal"    "lntamaño"    "fac500a"    
[31] "sch"         "exper"

In [ ]:
summary(enaho)

     pais                aÑo            mes             dpto      
 Length:45158       Min.   :2021   Min.   : 1.00   Min.   : 1.00  
 Class :character   1st Qu.:2021   1st Qu.: 4.00   1st Qu.: 8.00  
 Mode  :character   Median :2021   Median : 7.00   Median :14.00  
                    Mean   :2021   Mean   : 6.52   Mean   :13.29  
                    3rd Qu.:2021   3rd Qu.: 9.00   3rd Qu.:19.00  
                    Max.   :2021   Max.   :12.00   Max.   :25.00  
                                                                  
      zona           urban           conglome       vivienda        
 Min.   :1.000   Min.   :0.0000   Min.   : 5007   Length:45158      
 1st Qu.:1.000   1st Qu.:0.0000   1st Qu.:15034   Class :character  
 Median :2.000   Median :1.0000   Median :16854   Mode  :character  
 Mean   :2.143   Mean   :0.6879   Mean   :15328                     
 3rd Qu.:3.000   3rd Qu.:1.0000   3rd Qu.:18649                     
 Max.   :4.000   Max.   :1.0000   Max.   :21001   

## Omisión de Variables Relevantes

El primer modelo es 
\begin{gather*}
log(ingreso) = \beta_{1} + \beta_{2} sch + \epsilon
\end{gather*}

In [ ]:
ols_1 <- lm(lnwage ~ sch, data = enaho)
summary(ols_1)


Call:
lm(formula = lnwage ~ sch, data = enaho)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.4085 -0.5001  0.1618  0.6457  4.3800 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 5.775694   0.011913  484.83   <2e-16 ***
sch         0.091727   0.001081   84.84   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1 on 45156 degrees of freedom
Multiple R-squared:  0.1375,	Adjusted R-squared:  0.1375 
F-statistic:  7198 on 1 and 45156 DF,  p-value: < 2.2e-16


Sin embargo, alguien puede argumentar que la relación entre salarios y escolaridad está siendo impulsada por las diferencias de residencia. El segundo modelo es 
\begin{gather*}
log(ingreso) = \beta_{1} + \beta_{2} sch + \beta_{3} urban + \epsilon
\end{gather*}

In [ ]:
ols_2 <- lm(lnwage ~ sch + urban, data = enaho)
summary(ols_2)


Call:
lm(formula = lnwage ~ sch + urban, data = enaho)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.4668 -0.4804  0.1520  0.6400  4.4612 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 5.694484   0.012174  467.76   <2e-16 ***
sch         0.079223   0.001164   68.07   <2e-16 ***
urban       0.302029   0.010935   27.62   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.9919 on 45155 degrees of freedom
Multiple R-squared:  0.1518,	Adjusted R-squared:  0.1518 
F-statistic:  4041 on 2 and 45155 DF,  p-value: < 2.2e-16


In [ ]:
install.packages("memisc")
library("memisc")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Loading required package: lattice

Loading required package: MASS


Attaching package: ‘memisc’


The following objects are masked from ‘package:stats’:

    contr.sum, contr.treatment, contrasts


The following object is masked from ‘package:base’:

    as.array




In [ ]:
table_1 <- mtable("Model 1" = ols_1,
                  "Model 2" = ols_2,
            summary.stats = c("N", "R-squared"),
            coef.style = "default", 
            digits=3)
table_1

Calls:
Model 1: lm(formula = lnwage ~ sch, data = enaho)
Model 2: lm(formula = lnwage ~ sch + urban, data = enaho)

===========================================
                 Model 1       Model 2     
-------------------------------------------
  (Intercept)      5.776***      5.694***  
                  (0.012)       (0.012)    
  sch              0.092***      0.079***  
                  (0.001)       (0.001)    
  urban                          0.302***  
                                (0.011)    
-------------------------------------------
  N            45158         45158         
  R-squared        0.137         0.152     
===========================================
  Significance: *** = p < 0.001;   
                ** = p < 0.01;   
                * = p < 0.05

El retorno a la educación se reduce de 9.17% a 7.92% cuando se controla por la variable "urban". ¿Este coeficiente está sobreestimado? Ver que el coeficiente para sch en modelo 2 es menor que en modelo 1. ¿Por qué? Esto se relaciona con la formula de variable omitida. La variable "urban" es omitida en modelo 1. Por lo tanto: 

\begin{gather*}
E(\hat{\beta}_{2}|sch)=\beta_{2} + \beta_{3}\hat{\gamma}
\end{gather*}

donde $\beta_{3}$ es la relación entre ingresos y area, y $\hat{\gamma}$ es el coeficiente de la pendiente de la regresión de "urban" sobre "sch". Dado que individuos que viven en área urbana ganan en promedio más que los que viven en área rural $\beta_{3}>0$. Además, para obtener $\hat{\gamma}$, corremos el siguiente modelo:

In [ ]:
summary(lm(sch ~ urban, data = enaho))


Call:
lm(formula = sch ~ urban, data = enaho)

Residuals:
     Min       1Q   Median       3Q      Max 
-11.2624  -2.2624  -0.2624   3.3926  10.3926 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  7.60737    0.03379  225.17   <2e-16 ***
urban        3.65508    0.04073   89.73   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 4.011 on 45156 degrees of freedom
Multiple R-squared:  0.1513,	Adjusted R-squared:  0.1513 
F-statistic:  8052 on 1 and 45156 DF,  p-value: < 2.2e-16


El coeficiente de urban implica que, para la muestra analizada, individuos en área urbana en promedio son más educados que individuos que viven en área rural, por lo tanto $\hat{\gamma}>0$. Estas dos observaciones implica que $\hat{\beta}_{2}$ es sesgado hacia arriba en modelo 1.

## Usando operaciones matriciales

In [ ]:
X <- cbind(1,
          enaho$sch, 
          enaho$urban)
y <- enaho$lnwage

In [ ]:
b_hat <- solve(crossprod(X)) %*% crossprod(X, y)
b_hat

[,1]      
[1,] 5.69448441
[2,] 0.07922293
[3,] 0.30202883

In [ ]:
# Reportamos el sigma^2
y_hat <- X %*% b_hat
e_hat <- y - y_hat
sigma2_hat <- crossprod(e_hat) / (length(y) - ncol(X))
sigma2_hat

[,1]     
[1,] 0.9838437

In [ ]:
# Matriz de varianza covarianzas
V_hat <- drop(sigma2_hat) * solve(crossprod(X))
se <- sqrt(diag(V_hat))
se

[1] 0.012173900 0.001163833 0.010935304

## Inclusión de variables irrelevantes

In [ ]:
ols_3 <- lm(lnwage ~ sch + urban + conglome, data = enaho)
summary(ols_3)


Call:
lm(formula = lnwage ~ sch + urban + conglome, data = enaho)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.4625 -0.4802  0.1523  0.6396  4.4640 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  5.711e+00  1.986e-02 287.561   <2e-16 ***
sch          7.919e-02  1.164e-03  68.024   <2e-16 ***
urban        3.030e-01  1.097e-02  27.613   <2e-16 ***
conglome    -1.093e-06  1.045e-06  -1.046    0.296    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.9919 on 45154 degrees of freedom
Multiple R-squared:  0.1518,	Adjusted R-squared:  0.1518 
F-statistic:  2694 on 3 and 45154 DF,  p-value: < 2.2e-16


In [ ]:
table_2 <- mtable("Model 1" = ols_1,
                  "Model 2" = ols_2,
                  "Model 3" = ols_3,
            summary.stats = c("N", "R-squared"),
            coef.style = "default", 
            digits=8)
table_2

Calls:
Model 1: lm(formula = lnwage ~ sch, data = enaho)
Model 2: lm(formula = lnwage ~ sch + urban, data = enaho)
Model 3: lm(formula = lnwage ~ sch + urban + conglome, data = enaho)

========================================================================
                    Model 1            Model 2            Model 3       
------------------------------------------------------------------------
  (Intercept)      5.77569417***      5.69448441***      5.71089249***  
                  (0.01191273)       (0.01217390)       (0.01985978)    
  sch              0.09172738***      0.07922293***      0.07919287***  
                  (0.00108117)       (0.00116383)       (0.00116419)    
  urban                               0.30202883***      0.30296148***  
                                     (0.01093530)       (0.01097160)    
  conglome                                              -0.00000109     
                                                        (0.00000104)    
------------------------------------------------------------------------
  N            45158              45158              45158              
  R-squared        0.13748644         0.15181556         0.15183610     
========================================================================
  Significance: *** = p < 0.001; ** = p < 0.01; * = p < 0.05

Los errores estándar de los coeficientes aumentan. 

## Teorema Frisch-Waugh

Deseamos estudiar el impacto puro de la educación sobre los salario, entonces se debería limpiar el efecto del área. Se hace un procedimiento en tres etapas:

1. Regresionamos escolaridad contra área y calculamos los residuos. Estos residuos son aquella parte de la escolaridad que no es explicada por el área.

\begin{gather*}
sch = a_{1} + a_{2} urban + \epsilon
\end{gather*}

Los residuos estimados son:

\begin{gather*}
\hat{\epsilon}_{sch} = sch - \hat{a}_{1} + \hat{a}_{2} urban
\end{gather*}

In [ ]:
ols4 <- lm(sch ~ urban, data=enaho)
summary(ols4)


Call:
lm(formula = sch ~ urban, data = enaho)

Residuals:
     Min       1Q   Median       3Q      Max 
-11.2624  -2.2624  -0.2624   3.3926  10.3926 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  7.60737    0.03379  225.17   <2e-16 ***
urban        3.65508    0.04073   89.73   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 4.011 on 45156 degrees of freedom
Multiple R-squared:  0.1513,	Adjusted R-squared:  0.1513 
F-statistic:  8052 on 1 and 45156 DF,  p-value: < 2.2e-16


In [ ]:
# Calcular residuos
enaho$e_sch <- fitted(ols4) - enaho$sch

2. Regresionamos al lnwage contra el area y también calculamos los residuos de esta relación. De esta manera, se obtiene la variabilidad de los salarios sin influencia del área:

\begin{gather*}
log(ingreso) = b_{1} + b_{2} urban + v
\end{gather*}

Los residuos estimados son:

\begin{gather*}
\hat{\epsilon}_{lnwage} = lnwage - \hat{b}_{1} + \hat{b}_{2} urban
\end{gather*}

In [ ]:
ols5 <- lm(lnwage ~ urban, data=enaho)
summary(ols5)


Call:
lm(formula = lnwage ~ urban, data = enaho)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.3291 -0.5348  0.1266  0.6729  4.2751 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 6.297162   0.008774  717.73   <2e-16 ***
urban       0.591595   0.010578   55.93   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.042 on 45156 degrees of freedom
Multiple R-squared:  0.06478,	Adjusted R-squared:  0.06476 
F-statistic:  3128 on 1 and 45156 DF,  p-value: < 2.2e-16


In [ ]:
# Calcular residuos
enaho$e_lnwage <- fitted(ols5) - enaho$lnwage

3. Regresionamos a los residuos de los ingresos contra los residuos de la escolaridad en un modelo sin intercepto:

\begin{gather*}
\hat{\epsilon}_{lnwage} = g\hat{\epsilon}_{sch} + p
\end{gather*}

In [ ]:
summary(lm(e_lnwage ~ e_sch -1, data=enaho))


Call:
lm(formula = e_lnwage ~ e_sch - 1, data = enaho)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.4612 -0.6400 -0.1520  0.4804  6.4668 

Coefficients:
      Estimate Std. Error t value Pr(>|t|)    
e_sch 0.079223   0.001164   68.07   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.9919 on 45157 degrees of freedom
Multiple R-squared:  0.09307,	Adjusted R-squared:  0.09305 
F-statistic:  4634 on 1 and 45157 DF,  p-value: < 2.2e-16


El teorema de Frisch-Waugh afirma que el estimador $g$ es exactamente el mismo que el estimador de la pendiente de la sch obtenido por MCO del modelo: 
\begin{gather*}
log(ingreso) = \beta_{1} + \beta_{2} sch + \beta_{3} urban + \epsilon
\end{gather*}

In [ ]:
summary(lm(lnwage ~ sch + urban, data=enaho))


Call:
lm(formula = lnwage ~ sch + urban, data = enaho)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.4668 -0.4804  0.1520  0.6400  4.4612 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 5.694484   0.012174  467.76   <2e-16 ***
sch         0.079223   0.001164   68.07   <2e-16 ***
urban       0.302029   0.010935   27.62   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.9919 on 45155 degrees of freedom
Multiple R-squared:  0.1518,	Adjusted R-squared:  0.1518 
F-statistic:  4041 on 2 and 45155 DF,  p-value: < 2.2e-16


En conclusión, incorporar a "urban" como un regresor equivale a "limpiar" su efecto de las demás variables del modelo.